abe/abe.go

```diff
@@ -0,0 +1,472 @@
+func newAbe(l int) *Abe {
+	return &Abe{Params: &abeParams{
+		l: l, // number of attributes in the whole universe
+		p: bn256.Order,
+	}}
+}
+
+// AbePubKey represents a public key of an ABE scheme.
+type AbePubKey struct {
+	t data.VectorG2
+	y *bn256.GT
+}
+
+// GenerateMasterKeys generates a new set of public keys, needed
+// for encrypting data, and secret keys needed for generating keys
+// for decryption.
+func (a Abe) GenerateMasterKeys() (*AbePubKey, data.Vector, error) {
+	// todo: should we put public key into Abe struct?
+	sampler := sample.NewUniform(a.Params.p)
+	sk, err := data.NewRandomVector(a.Params.l+1, sampler)
+	if err != nil {
+		return nil, nil, err
+	}
+	t := sk[:a.Params.l].MulG2()
+	y := new(bn256.GT).ScalarBaseMult(sk[a.Params.l])
+
+	return &AbePubKey{t: t, y: y}, sk, nil
+}
+
+// AbeCipher represents a ciphertext of an ABE scheme.
+type AbeCipher struct {
+	gamma     []int         // the set of attributes that can be used for policy of decryption
+	attribToI map[int]int   // a map that connects the attributes in gamma with elements of e
+	e0        *bn256.GT     // the first part of the encryption
+	e         data.VectorG2 // the second part of the encryption
+}
+
+// Encrypt takes as an input a message msg represented as an element of an elliptic
+// curve, gamma a set of attributes that can be latter used to in the decryption policy
+// and a public key pk. It returns an encryption of msk. An case of a failed procedure
+// an error is returned.
+func (a Abe) Encrypt(msg *bn256.GT, gamma []int, pk *AbePubKey) (*AbeCipher, error) {
+	sampler := sample.NewUniform(a.Params.p)
+	s, err := sampler.Sample()
+	if err != nil {
+		return nil, err
+	}
+
+	e0 := new(bn256.GT).Add(msg, new(bn256.GT).ScalarMult(pk.y, s))
+	e := make(data.VectorG2, len(gamma))
+	attribToI := make(map[int]int)
+	for i, el := range gamma {
+		e[i] = new(bn256.G2).ScalarMult(pk.t[el], s)
+		attribToI[el] = i
+	}
+
+	return &AbeCipher{gamma: gamma,
+		attribToI: attribToI,
+		e0:        e0,
+		e:         e}, nil
+}
+
+// Msp represents a monotone span program (MSP) describing a policy which
+// attributes are needed to decrypt the ciphertext. It includes a matrix
+// mat and a mapping from the rows of the mat to attributes. A MSP policy
+// allows decryption of an entity with a set of attributes A iff all the
+// rows of the matrix mapped to an element of A span the vector [1, 1,..., 1]
+// in Z_p.
+type Msp struct {
+	p           *big.Int
+	mat         data.Matrix
+	rowToAttrib []int
+}
+
+// KeyGen given a monotone span program (MSP) msp and the vector of secret
+// keys produces a vector of keys needed for the decryption. In particular,
+// for each row of the MSP matrix msp.mat it creates a corresponding key. Since
+// each row of msp.mat has a corresponding key, this keys can be latter delegated
+// to entities with corresponding attributes.
+func (a Abe) KeyGen(msp *Msp, sk data.Vector) (data.VectorG1, error) {
```

Comment:
Rename the function to something more descriptive. Something like `GenerateTypeKey`, where `Type` is the type of key, maybe?

Reply:
I renamed it into GeneratePolicyKeys.